### PR TITLE
CN0401 demo application

### PR DESCRIPTION
**Pull request by adrimbarean**
This patch adds the initial commit of the CN0401 ADICUP3029 project, SPI to CAN translator. The patch features:
 - minimal API for the CAN FD controller;
 - application code including a Command Line Interface that can be used to connect 2 or more systems as nodes into a CAN bus.

### CODE REVIEW

**Comment by adrimbarean** [comment]
V0 -> V1:
 - fix AStyle errors;
 - fix //todo in code.

**Comment by adrimbarean** [comment]
V1 -> V2:
 - fix more AStyle and //todo comments;
 - rebase on master to include the Travis fix.

**Comment by adrimbarean** [comment]
V2 -> V3:
 - change the production test command from "test" to "t".

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 1]
```diff
@@ -0,0 +1,1353 @@
...
+ * @param [in] data    - New value of the byte.
+ *
+ * @return 0 in case of success, error code in case of failure.
+ */
+int32_t can_ctrl_sfr_byte_write(struct can_ctrl_dev *dev, uint16_t address,
+				uint8_t data)
+{
+	uint8_t spi_msg[] = {0, 0, 0};
+
+	spi_msg[0] |= CAN_CTRL_CMD_MODE(CAN_CTRL_CMD_WRITE);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: these two can be done in one go:
```
spi_msg[0] = CAN_CTRL_CMD_MODE(CAN_CTRL_CMD_WRITE) |
CAN_CTRL_ADDR_UP_NIBBLE_MODE(address);
```
Also for the others you can remove the OR operator.

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 2]
```diff
@@ -0,0 +1,1353 @@
...
+ *
+ * @return 0 in case of success, error code in case of failure.
+ */
+int32_t can_ctrl_ram_read(struct can_ctrl_dev *dev, uint16_t address,
+			  uint32_t *data)
+{
+	uint8_t spi_msg[] = {0, 0, 0, 0, 0, 0};
+	int32_t ret;
+
+	spi_msg[0] |= CAN_CTRL_CMD_MODE(CAN_CTRL_CMD_READ);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: same as above

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 3]
```diff
@@ -0,0 +1,1353 @@
...
+		{0,  1, 0, 0,  2}  /* 10M  */
+	};
+	int32_t ret;
+	uint32_t temp;
+
+	/* Nominal bit rate */
+	ret = can_ctrl_sfr_word_read(dev, CAN_CTRL_REG_NBTCFG, &temp);
+	if(ret != 0)
+		return ret;
+	temp &= ~NBTCFG_BRP_MODE(WORD_MASK);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: these can be reduced

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 4]
```diff
@@ -0,0 +1,1353 @@
...
+		}
+	}
+
+	obj.word[0] = msg_buff[1];
+	obj.word[1] = msg_buff[0];
+	payload_size = can_ctrl_dlc_to_size(obj.hdr_str.dlc);
+
+	if(fifo_con & FIFOCON_RXTSEN_MASK)
+		for(i = 0; i < payload_size; i++)
+			data[i] = ((uint8_t *)msg_buff)[i + 12];
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: memcpy ?

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 5]
```diff
@@ -0,0 +1,1353 @@
...
+ */
+static int32_t can_ctrl_gpio_config(struct can_ctrl_dev *dev)
+{
+	int32_t ret;
+	uint32_t temp;
+
+	ret = can_ctrl_sfr_word_read(dev, CAN_CTRL_REG_IOCON, &temp);
+	if(ret != 0)
+		return ret;
+	temp &= ~IOCON_PM1_MODE(WORD_MASK);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: this can be reduced.

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 6]
```diff
@@ -0,0 +1,1353 @@
...
+static int32_t can_ctrl_con_reg_config(struct can_ctrl_dev *dev,
+				       struct can_ctrl_init_param *init_param)
+{
+	int32_t ret;
+	uint32_t temp;
+
+	ret = can_ctrl_sfr_word_read(dev, CAN_CTRL_REG_CON, &temp);
+	if(ret != 0)
+		return ret;
+	temp &= ~CON_TXQEN_MODE(WORD_MASK);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: same as above

File: projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c [hunk 7]
```diff
@@ -0,0 +1,1353 @@
...
+	flags >>= 1;
+	temp |= INT_RXOVIE_MODE(flags);
+	flags >>= 1;
+	temp |= INT_SERRIE_MODE(flags);
+	flags >>= 1;
+	temp |= INT_CERRIE_MODE(flags);
+	flags >>= 1;
+	temp |= INT_WAKIE_MODE(flags);
+	flags >>= 1;
+	temp |= INT_IVMIE_MODE(flags);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c]: to reduce

File: projects/ADuCM3029_demo_cn0401/src/cn0401.c [hunk 8]
```diff
@@ -0,0 +1,758 @@
...
+			      (uint8_t *)"#        #   #   #  #       ### ###\n");
+	usr_uart_write_string(dev->cli_handler->uart_device,
+			      (uint8_t *)"#####   #     #  #  #        #   #  \n");
+	usr_uart_write_string(dev->cli_handler->uart_device,
+			      (uint8_t *)"#       #######  #  #               \n");
+	usr_uart_write_string(dev->cli_handler->uart_device,
+			      (uint8_t *)"#       #     #  #  #       ### ### \n");
+	usr_uart_write_string(dev->cli_handler->uart_device,
+			      (uint8_t *)"#       #     # ### ####### ### ### \n\n\n");
+
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0401/src/cn0401.c]: nitpick: extra line

**Comment by adrimbarean** [comment]
V3 -> V4:
 - remove extra line;
 - reduce clutter code.

**Pull request merged**